### PR TITLE
Support the b64 header parameter

### LINKED ISSUE #225: Support the "b64" header parameter
https://tools.ietf.org/html/rfc7797#section-3

Currently there is no way to sign the payload *not* encoded in base64. That's a serious limitation when the service you're sending the signatures to requires `b64=false`. If `SignerOptions.ExtraHeaders` has `b64` key with a `false` value then the payload should not be encoded.

### PR DESCRIPTION
**Pull request by philtay**
Closes #225

### CODE REVIEW

**Comment by philtay** [comment]
This PR is ready for review.

/cc @csstaub @mcpherrinm

File: jws.go [hunk 1]
```diff
@@ -94,19 +95,38 @@ func (sig Signature) mergedHeaders() rawHeader {
 
 // Compute data to be signed
 func (obj JSONWebSignature) computeAuthData(payload []byte, signature *Signature) []byte {
-	var serializedProtected string
+	var buf bytes.Buffer
+
+	pro := new(rawHeader)
```
Comment (csstaub) [inline jws.go]: This is a nit, but I'd prefer if we gave these variables more descriptive names. It's not clear at first glance what "pro", "enc" and "buf" mean or what they're used for.

Comment (csstaub) [reply jws.go]: How about calling these protectedHeader, needsBase64, and authData (or something along those lines)?

File: shared.go [hunk 2]
```diff
@@ -349,6 +350,21 @@ func (parsed rawHeader) getP2S() (*byteBuffer, error) {
 	return parsed.getByteBuffer(headerP2S)
 }
 
+// getB64 extracts parsed "b64" from the raw JSON, defaulting to true.
+func (parsed rawHeader) getB64() (bool, error) {
```
Comment (csstaub) [inline shared.go]: Would it ever be necessary here to distinguish between the header being present, and set, or the header being not present and the default return value being false?

File: signing.go [hunk 3]
```diff
@@ -254,11 +255,26 @@ func (ctx *genericSigner) Sign(payload []byte) (*JSONWebSignature, error) {
...
-		input := []byte(fmt.Sprintf("%s.%s",
-			base64.RawURLEncoding.EncodeToString(serializedProtected),
-			base64.RawURLEncoding.EncodeToString(payload)))
+		var input bytes.Buffer
 
-		signatureInfo, err := recipient.signer.signPayload(input, recipient.sigAlg)
+		input.WriteString(base64.RawURLEncoding.EncodeToString(serializedProtected))
+		input.WriteByte('.')
+
+		enc := true
```
Comment (csstaub) [inline signing.go]: Same comment as above, let's call this something other than "enc". Could be confused to mean "encrypt" for example.

File: signing_test.go [hunk 4]
```diff
@@ -521,3 +521,39 @@ func TestSignerExtraHeaderInclusion(t *testing.T) {
...
 }
+
+func TestSignerB64(t *testing.T) {
+	signingKey, verificationKey := GenerateSigningTestKey(RS256)
+
+	opts := &SignerOptions{
+		ExtraHeaders: map[HeaderKey]interface{}{
+			"b64": false,
+		},
+	}
```
Comment (csstaub) [inline signing_test.go]: This seems somewhat non-obvious, would it be better to have this be a first-class option in SignerOptions and have the signing code add the header if that's selected?

**Comment by csstaub** [comment]
Hi @philtay, thank you so much for your pull request! I left a couple of comments inline, but I think what's also missing is code to add the b64 header to the list of crit headers (see RFC 7797, Section 6: https://tools.ietf.org/html/rfc7797#section-6).

**Comment by philtay** [comment]
@csstaub thanks for the review. The last commit takes in account your comments and brings some improvements.

**Comment by philtay** [comment]
@csstaub can you review the updated PR? I'm eager to have this merged. Thanks!

**Comment by csstaub** [comment]
Hi @philtay, changes look good to me! Thank you very much for your contribution. I'll merge this and play around with it a bit and then will try to make a tagged release later this week.

**Comment by philtay** [comment]
@csstaub thanks! I've another PR to open before making a tagged release. A `func (obj JSONWebSignature) DetachedCompactSerialize() (string, error)` method as per RFC 7797 (compact serialization and detached payload). I'll open the PR later today or tomorrow. If you want, I'm also interested in closing #223 and cutting a v3.0.0. Just merge the `v2` branch  into `master` and I'll do the rest. Let me know.

**Comment by csstaub** [comment]
Sounds good, let's do it.

**Pull request merged**